src/core/meshcop/commissioner.cpp

```diff
@@ -968,16 +958,12 @@ void Commissioner::SendKeepAlive(uint16_t aSessionId)
     otLogInfoMeshCoP("sent keep alive");
 
 exit:
-
-    if (error != OT_ERROR_NONE)
+    if ((error != OT_ERROR_NONE) && (message != nullptr))
     {
-        otLogWarnMeshCoP("Failed to send keep alive: %s", otThreadErrorToString(error));
-
-        if (message != nullptr)
-        {
-            message->Free();
-        }
+        message->Free();
     }
+
+    LogError("send keep alive", error);
```

Comment:
```suggestion
    LogError("send", "keep alive", error);
```
Should we split `action` and `message type` ? The actions can be `send`, `start`, `process`, etc.

Reply:
Having more function parameters would increase code size on caller side (now requiring two arguments to be prepared (loaded into registers, etc.) before method/function call). The code-size reduction from potential sharing of strings "send", "start", etc would need to be balanced by this. 

Also the sharing of common string definitions is not always done by all toolchains. For example,  some compilers (particularly arm architecture) can prefer (under certain optimization flags) to include all constant (arrays/strings/pointers) in the region directly at the end of each method so that instructions that load relative to `PC` (program counter) register can be used to access/load them quickly.

If the goal is to reduce code size by sharing string, it would help if we can do this explicitly (e.g. , define an `Action` enum to covert to string within `LogError` itself) and/or share the message type (`"keep alive"`) strings as well.

 I would propose to leave this for a separate PR (future change)

Reply:
> Having more function parameters would increase code size on caller side (now requiring two arguments to be prepared (loaded into registers, etc.) before method/function call). The code-size reduction from potential sharing of strings "send", "start", etc would need to be balanced by this.

We can easily compensate for the code size increase for one more parameter if the action strings are reused? Especially when the action strings are not only shared in meshcop, but also reused for other modules (e.g. MLE). 

> Also the sharing of common string definitions is not always done by all toolchains. For example, some compilers (particularly arm architecture) can prefer (under certain optimization flags) to include all constant (arrays/strings/pointers) in the region directly at the end of each method so that instructions that load relative to `PC` (program counter) register can be used to access/load them quickly.

But since `-Os` is used, compilers should prefer smaller code than than higher performance? I think we can have the benefit of reduced code size on majority of compilers (at least GCC) even without using string enum. The benefit should already overrun the potential risk. 

> If the goal is to reduce code size by sharing string, it would help if we can do this explicitly (e.g. , define an `Action` enum to covert to string within `LogError` itself) and/or share the message type (`"keep alive"`) strings as well.
> 
> I would propose to leave this for a separate PR (future change)

I think we should do it. Fine to me to leave this for a separate PR.

Reply:
> We can easily compensate for the code size increase for one more parameter if the action strings are reused? Especially when the action strings are not only shared in meshcop, but also reused for other modules (e.g. MLE).

In general yes, but not necessarily in all cases. There are counter-intuitive cases.
It depends on different factors, e.g., size of strings being shared, how the new arg is added. 

Let me give an example:

If we share `"send"` string,  we would save 4 chars or 4 bytes (per its use). But adding a new parameter/argument in the `LogError()` call can add more than 4 bytes (6 bytes or more on arm-gcc) to the generated code (again per use).

Here is disassembly of code with two args in `LogError()` like `LogError("send", "scan results", error)`
(this is from arm-gcc-7 and same config as the one we use in our size-report)

```
# EnegyScanServer::SendReport()

00011868 <_ZN2ot16EnergyScanServer10SendReportEv>:
   11868:       b5f0            push    {r4, r5, r6, r7, lr}
   ....

   # Loading the two pointer values into r0 and r1 registers 
   # before branching to `MeshCoP::LogError`

   1194c:       490a            ldr     r1, [pc, #40]   ; (11978 <_ZN2ot16EnergyScanServer10SendReportEv+0x110>)
   1194e:       480b            ldr     r0, [pc, #44]   ; (1197c <_ZN2ot16EnergyScanServer10SendReportEv+0x114>)
   11950:       f7fd fa5a       bl      ee08 <_ZN2ot7MeshCoP8LogErrorEPKcS2_7otError>
      ...
   11966:       e7f0            b.n     1194a <_ZN2ot16EnergyScanServer10SendReportEv+0xe2>
   11968:       2000a344        .word   0x2000a344
   1196c:       00046403        .word   0x00046403
   11970:       2000ac2c        .word   0x2000ac2c
   11974:       00045694        .word   0x00045694
   11978:       000456aa        .word   0x000456aa
   1197c:       00045930        .word   0x00045930
```

Note that we see two pointers being loaded into `r0` and `r1` before calling `LogError()`. We need two bytes for `ldr  r1, [pc, #40]` instruction for the second argument, but in addition to this, the actual pointer value is stored right at the end of function at `[pc, #40]` or at address `11978` (which stores the string address `00456aa`). This adds another 4 bytes to the code. 

---

Another good (counter-intuitive) example is our logging:
- We have the feature to append log level and/or log region prefix string to the log.
- The prefix string are literally added to the `format` string, so the strings (in the image) are like:
  - `"[NOTE]-MLE-----: RLOC16 %04x -> %04x"`, or `"[NOTE]-MLE-----: Downgrade to REED"`
- At first it seems to be inefficient and makes sense to change it to reuse the `"[NOTE]"` and `"-MLE----"` strings.
- We tried this, which boils down to a call like this:
   - `otPlatLog("%s%s: " aFormat, sLevelPrefix[aLevel], sRegionPrefix[aRegion], ...)`
   - `sLevelPrefix[]` is const array of strings (for each log level), same way for `sRegionPrefix[]` string array.
- This change actually increased the code size largely (by ~2K if I recall).
- The reason is the overhead of adding two arguments to `otPlatLog()` call (which is variable arg function itself) was more than the bytes/chars saved by sharing the string literals.

-----

> But since -Os is used, compilers should prefer smaller code than than higher performance? I think we can have the benefit of reduced code size on majority of compilers (at least GCC) even without using string enum. The benefit should already overrun the potential risk.

Over the years, I have been surprised many times by toolchain behaviors (specially in embedded sw where we have many custom/specialized compilers). My approach/advice is that it is always better not to rely on toolchains and/or assume any behavior. Better to be explicit in the code (when/if we can).

Reply:
Thanks for the explanation.